File: 2023/Day1/solve.cpp
#include <iostream>
#include <fstream>
#include <string>
#include <unordered_map>

void part1() {
    std::ifstream input("input.txt");

    int sum = 0;
    while(!input.eof()) {
        std::string line;
        std::getline(input, line);
        int i = 0, j = line.length() - 1;
        while(!isdigit(line[i])) i++;
        while(!isdigit(line[j])) j--;
        sum += (line[i] - '0') * 10 + (line[j] - '0');
    }

    std::cout << "Part1: " << sum << std::endl;
}

void replace(std::string &str, const std::string &from, const std::string &to) {
    size_t start_pos = str.find(from);
    if (start_pos == std::string::npos) {
        return;
    }
    str.replace(start_pos, from.length(), to);
}

void part2() {
    std::ifstream input("input.txt");
    const std::unordered_map<std::string, int> dict{
        {"one", 1},
        {"two", 2},
        {"three", 3},
        {"four", 4},
        {"five", 5},
        {"six", 6},
        {"seven", 7},
        {"eight", 8},
        {"nine", 9}};

    int sum = 0;
    while (!input.eof()) {
        std::string line;
        std::getline(input, line);

        int first = -1, last = 0;
        for (int i = 0; i < line.length(); i++) {
            if (isdigit(line[i])) {
                if (first == -1) {
                    first = line[i] - '0';
                }
                last = line[i] - '0';
            } else {
                for (auto &[key, value] : dict) {
                    if (line.substr(i, key.length()) == key) {
                        if (first == -1) {
                            first = value;
                        }
                        last = value;
                    }
                }
            }
        }
        sum += first * 10 + last;
    }

    std::cout << "Part2: " << sum << std::endl;
}

int main() {
    part1();
    part2();
    return 0;
}

File: 2023/Day1/solve.rs
use std::fs;
use std::collections::HashMap;

fn part1() {
    let input = fs::read_to_string("input.txt").unwrap();

    let mut sum = 0;
    for line in input.lines() {
        let first = line.chars().filter(|c| c.is_digit(10)).next().unwrap();
        let last = line.chars().filter(|c| c.is_digit(10)).last().unwrap();

        sum += first.to_digit(10).unwrap() * 10 + last.to_digit(10).unwrap();
    }

    println!("Sum: {}", sum);
}

fn part2() {
    let input = fs::read_to_string("input.txt").unwrap();
    let mut dict = HashMap::new();

    dict.insert(String::from("one"), 1);
    dict.insert(String::from("two"), 2);
    dict.insert(String::from("three"), 3);
    dict.insert(String::from("four"), 4);
    dict.insert(String::from("five"), 5);
    dict.insert(String::from("six"), 6);
    dict.insert(String::from("seven"), 7);
    dict.insert(String::from("eight"), 8);
    dict.insert(String::from("nine"), 9);

    let mut sum = 0;
    for line in input.lines() {
        let mut first = -1;
        let mut last = 0;

        for (i, c) in line.chars().enumerate() {
            if c.is_digit(10) {
                if first == -1 {
                    first = c.to_digit(10).unwrap() as i32;
                }
                last = c.to_digit(10).unwrap() as i32;
            } else {
                for (key, val) in &dict {
                    if line[i..].starts_with(key) {
                        if first == -1 {
                            first = *val;
                        }
                        last = *val;
                    }
                }
            }
        }

        sum += first * 10 + last;
    }

    println!("{}", sum);
}

fn main() {
    part1();
    part2();
}